securebanking-openbanking-uk-gateway-conversion-filter/src/main/java/com/forgerock/securebanking/uk/gateway/conversion/filter/IntentConverterFilter.java

```diff
@@ -0,0 +1,211 @@
+
+    private final IntentType intentType;
+    public static final String CONFIG_FIELD_INTENT_TYPE = "intentType";
+    private final MessageType payloadFrom;
+    public static final String CONFIG_FIELD_PAYLOAD_FROM = "payloadFrom";
+    private final List<MessageType> resultTo;
+    public static final String CONFIG_FIELD_RESULT_TO = "resultTo";
+
+    private static final ObjectMapper MAPPER = GenericConverterMapper.getMapper();
+
+    public IntentConverterFilter(final IntentType intentType, final MessageType payloadFrom) {
+        this(intentType, payloadFrom, List.of(MessageType.RESPONSE));
+    }
+
+    public IntentConverterFilter(final IntentType intentType, final MessageType payloadFrom, final List<MessageType> resultTo) {
+        this.intentType = intentType;
+        this.payloadFrom = payloadFrom;
+        this.resultTo = resultTo;
+    }
+
+    @Override
+    public Promise<Response, NeverThrowsException> filter(Context context, Request request, Handler next) {
+        try {
+            if (logger.isInfoEnabled()) {
+                printFilterInfo();
+            }
+            String jsonPayload = getEntity(context, request, next);
+
+            Object objectMapped = convert(jsonPayload);
+
+            byte[] serialised = toBytes(objectMapped);
+
+            return processResult(serialised, context, request, next);
+        } catch (Exception e) {
+            logger.error("Conversion to OB Object filter Error\n", e);
+            ResponseException responseException = new ResponseException(e.getMessage(), e);
+            // Overriding the internal server error status to bad request
+            responseException.getResponse().setStatus(Status.BAD_REQUEST);
+            return newResponsePromise(responseException.getResponse());
+        }
+    }
+
+    /** Creates and initializes a ConversionFilter in a heap environment. */
+    public static class Heaplet extends GenericHeaplet {
+
+        @Override
+        public Object create() {
+            IntentType intentType = config.get(CONFIG_FIELD_INTENT_TYPE)
+                    .required()
+                    .as(enumConstant(IntentType.class));
+            logger.trace("intentType = {}", intentType);
+            MessageType payloadFrom = config.get(CONFIG_FIELD_PAYLOAD_FROM)
+                    .required()
+                    .defaultTo(MessageType.REQUEST.toString())
+                    .as(enumConstant(MessageType.class));
+            logger.trace("payload From {}", payloadFrom);
+            List<MessageType> payloadTo = config.get(CONFIG_FIELD_RESULT_TO)
+                    .defaultTo(List.of(MessageType.RESPONSE.toString()))
+                    .as(messageTypeList());
+            return new IntentConverterFilter(intentType, payloadFrom, payloadTo);
+        }
+    }
+
+    private static Function<JsonValue, List<MessageType>, JsonValueException> messageTypeList() {
```

Comment:
I think this is only called by the Heaplet, if so then move it inside the Heaplet class.

As the Heaplet logic is getting larger, I think it would make the IntentConverterFilter more readable if the Heaplet is converted from an inner class to a top level class in its own file